src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java

```diff
@@ -0,0 +1,487 @@
+
+        Logger.recordOutput("Odometry/Received Update From Module " + moduleIndex, true);
+        moduleDeltas[moduleIndex] =
+            new SwerveModulePosition(
+                modulePositions[moduleIndex].distanceMeters
+                    - lastModulePositions[moduleIndex].distanceMeters,
+                modulePositions[moduleIndex].angle);
+        lastModulePositions[moduleIndex] = modulePositions[moduleIndex];
+      }
+
+      // missing some data :(
+      if (hasNullModulePosition) {
+        missingModuleData.set(true);
+        if (!gyroInputs.connected
+            || sample.values().get(new SignalID(SignalType.GYRO, OdometryThreadIO.GYRO_MODULE_ID))
+                == null) {
+          missingGyroData.set(true);
+          // no modules and no gyro so we're just sad :(
+        } else {
+          missingGyroData.set(false);
+          // null here is checked by if clause
+          rawGyroRotation =
+              Rotation2d.fromDegrees(sample.values().get(new SignalID(SignalType.GYRO, -1)));
+          lastGyroRotation = rawGyroRotation;
+          Logger.recordOutput("Odometry/Gyro Rotation", lastGyroRotation);
+          Tracer.trace(
+              "update estimator",
+              () ->
+                  estimator.updateWithTime(
+                      sample.timestamp(), rawGyroRotation, lastModulePositions));
+        }
+        continue;
+      }
+
+      // If we have all our module data . . .
+      // The twist represents the motion of the robot since the last
+      // sample in x, y, and theta based only on the modules, without
+      // the gyro. The gyro is always disconnected in simulation.
+      Twist2d twist = kinematics.toTwist2d(moduleDeltas);
+      if (!gyroInputs.connected
+          || sample.values().get(new SignalID(SignalType.GYRO, OdometryThreadIO.GYRO_MODULE_ID))
+              == null) {
+        missingGyroData.set(true);
+        // We don't have a complete set of data, so just use the module rotations
+        rawGyroRotation = rawGyroRotation.plus(new Rotation2d(twist.dtheta));
+      } else {
+        missingGyroData.set(false);
+        rawGyroRotation =
+            Rotation2d.fromDegrees(
+                sample
+                    .values()
+                    .get(new SignalID(SignalType.GYRO, OdometryThreadIO.GYRO_MODULE_ID)));
+        twist =
+            new Twist2d(twist.dx, twist.dy, rawGyroRotation.minus(lastGyroRotation).getRadians());
+      }
+      // Apply the twist (change since last sample) to the current pose
+      lastGyroRotation = rawGyroRotation;
+      Logger.recordOutput("Odometry/Gyro Rotation", lastGyroRotation);
+      // Apply update
+      estimator.updateWithTime(sample.timestamp(), rawGyroRotation, modulePositions);
+    }
+  }
+
+  /**
+   * Generates a set of samples without using the async thread. Makes lots of Objects, so be careful
+   * when using it irl!
+   */
```

Comment:
could you instead like modify the previous `signalID`s somehow or am i yapping

Reply:
probably? given that its only used in sim i didnt care too much abt perf.